src/derive.rs

```diff
@@ -273,25 +273,31 @@ pub trait Subcommand: FromArgMatches + Sized {
 ///     level: Level,
 /// }
 ///
-/// #[derive(clap::ArgEnum)]
+/// #[derive(clap::ArgEnum, Clone)]
 /// enum Level {
 ///     Debug,
 ///     Info,
 ///     Warning,
 ///     Error,
 /// }
 /// ```
-pub trait ArgEnum: Sized {
-    /// All possible argument choices, in display order.
-    const VARIANTS: &'static [&'static str];
+pub trait ArgEnum: Sized + Clone {
+    /// All possible argument values, in display order.
+    fn value_variants<'a>() -> &'a [Self];
 
     /// Parse an argument into `Self`.
-    fn from_str(input: &str, case_insensitive: bool) -> Result<Self, String>;
+    fn from_str(input: &str, case_insensitive: bool) -> Result<Self, String> {
+        Self::value_variants()
+            .iter()
+            .find(|v| v.arg_value().unwrap().matches(input, case_insensitive))
```

Comment:
Bad `unwrap`? Should be handled correctly

Reply:
This is correct.  `value_variants` should only return variants that have an associated `ArgValue`, so `arg_value` will always return `Some(_)` in this call.

The `.unwrap()` could be replaced by an `except()`.  I can go either way; my general rule of thumb has been how close and obvious are the two sides of the contract.  The less obvious, the more clear you need to be.  I would have assumed this would have been obvious but I just got proven wrong.